src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java

```diff
@@ -0,0 +1,69 @@
+package org.broadinstitute.hellbender.tools.walkers.genotyper;
+
+import htsjdk.variant.variantcontext.Allele;
+import htsjdk.variant.variantcontext.GenotypeLikelihoods;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.genotyper.AlleleList;
+import org.broadinstitute.hellbender.utils.genotyper.AlleleListPermutation;
+import org.broadinstitute.hellbender.utils.genotyper.LikelihoodMatrix;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class delegates genotyping to allele count- and ploidy-dependent {@link GenotypeLikelihoodCalculator}s
+ * under the assumption that sample genotypes are independent conditional on their population frequencies.
+ */
+public final class IndependentSampleGenotypesModel {
+
+    private final int cacheAlleleCountCapacity;
+    private GenotypeLikelihoodCalculator[][] likelihoodCalculators;
+
+    public IndependentSampleGenotypesModel() { this(10,50); }
+
+    public IndependentSampleGenotypesModel(final int calculatorCachePloidyCapacity, final int calculatorCacheAlleleCapacity) {
+        cacheAlleleCountCapacity = calculatorCacheAlleleCapacity;
+        likelihoodCalculators = new GenotypeLikelihoodCalculator[calculatorCachePloidyCapacity][calculatorCacheAlleleCapacity];
+    }
+
+    public <A extends Allele> GenotypingLikelihoods<A> calculateLikelihoods(final AlleleList<A> genotypingAlleles, final GenotypingData<A> data) {
+        Utils.nonNull(genotypingAlleles, "the allele cannot be null");
+        Utils.nonNull(data, "the genotyping data cannot be null");
+
+        final AlleleListPermutation<A> permutation = data.permutation(genotypingAlleles);
+        final AlleleLikelihoodMatrixMapper<A> alleleLikelihoodMatrixMapper = AlleleLikelihoodMatrixMapper.newInstance(permutation);
+
+        final int sampleCount = data.numberOfSamples();
+        final PloidyModel ploidyModel = data.ploidyModel();
+        final List<GenotypeLikelihoods> genotypeLikelihoods = new ArrayList<>(sampleCount);
+        final int alleleCount = genotypingAlleles.numberOfAlleles();
+
+        GenotypeLikelihoodCalculator likelihoodsCalculator = sampleCount > 0 ? getLikelihoodsCalculator(ploidyModel.samplePloidy(0), alleleCount) : null;
+        for (int i = 0; i < sampleCount; i++) {
```

Comment:
Do we have good, independent unit tests covering the four cases you collapsed into a single method (0 samples, 1 sample, multi-sample heterogeneous, and multi-sample homogeneous)?